src/cls-ah.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const asyncHook = require('async_hooks');
+
+const wrappedSymbol = Symbol('context_wrapped');
+const contexts = {};
+var current = {};
+
+const hook = asyncHook.createHook({init, before, after, destroy});
+hook.enable();
+
+function Namespace() {}
+
+Namespace.prototype.get = function(k) {
+  return current ? current[k] : null;
+};
+
+Namespace.prototype.set = function(k, v) {
+  if (current) {
+    current[k] = v;
+  }
+};
+
+Namespace.prototype.run = function(fn) {
+  const oldContext = current;
+  current = {};
+  const res = fn();
+  current = oldContext;
+  return res;
+};
+
+Namespace.prototype.runAndReturn = function(fn) {
+  return this.run(fn);
+};
+
+Namespace.prototype.bind = function(cb) {
+  if (cb[wrappedSymbol] || !current) {
+    return cb;
+  }
+  const boundContext = current;
+  const wrapped = cb.length === 0 ? function() {
+    const oldContext = current;
+    current = boundContext;
+    const res = cb.apply(this, arguments);
+    current = oldContext;
+    return res;
+  } : function(a) {
+    const oldContext = current;
+    current = boundContext;
+    const res = cb.apply(this, arguments);
+    current = oldContext;
+    return res;
+  };
+  wrapped[wrappedSymbol] = true;
+  return wrapped;
+};
+
+function patchEventEmitterMethod(ee, method, ns) {
+  const oldMethod = ee[method];
+  ee[method] = function(e, f) { return oldMethod.call(this, e, ns.bind(f)); };
+}
+
+const eventEmitterMethods =
+  [ 'addListener', 'on', 'once', 'prependListener', 'prependOncelistener' ];
+
+Namespace.prototype.bindEmitter = function(ee) {
+  for (var method of eventEmitterMethods) {
+    patchEventEmitterMethod(ee, method, this);
+  }
+};
+
+var n = null;
+
+// AsyncWrap Hooks
+
+function init(uid, provider, parentUid, parentHandle) {
+  contexts[uid] = current;
+}
+
+function before(uid) {
+  current = contexts[uid];
+}
+
+function after(uid) {}
+
+function destroy(uid) {
+  delete contexts[uid];
+}
+
+module.exports = {
+  createNamespace: function() {
+    n = new Namespace();
```

Comment:
Given that `n` is a singleton, you don't expect `createNamespace` to be called more than once, correct? If that is the case, assert rather than clobbering.